src/Math-AutomaticDifferenciation/PMHyperDualNumber.class.st

```diff
@@ -66,160 +66,195 @@ PMHyperDualNumber >> / aHyperDualNumber [
 
 { #category : #'mathematical functions' }
 PMHyperDualNumber >> arcCos [
-|a d|
-a:=1 - value squared.
-d:=a sqrt negated.
-^(super arcCos) eps2: eps2 / d; eps1eps2: eps1eps2 /d + (eps * eps2 * value negated /(a raisedTo: (3/2)));yourself
-
+	| a d |
+	a := 1 - value squared.
+	d := a sqrt negated.
+	^ super arcCos
+		eps2: eps2 / d;
+		eps1eps2: eps1eps2 / d + (eps * eps2 * value negated / (a raisedTo: 3 / 2));
+		yourself
 ]
 
 { #category : #'mathematical functions' }
 PMHyperDualNumber >> arcSin [
-|a d|
-a:=1 - value squared.
-d:=a sqrt.
-^(super arcSin) eps2: eps2 / d; eps1eps2: eps1eps2 /d + (eps * eps2 * value /(a raisedTo: (3/2)));yourself
-
+	| a d |
+	a := 1 - value squared.
+	d := a sqrt.
+	^ super arcSin
+		eps2: eps2 / d;
+		eps1eps2: eps1eps2 / d + (eps * eps2 * value / (a raisedTo: 3 / 2));
+		yourself
 ]
 
 { #category : #'mathematical functions' }
 PMHyperDualNumber >> arcTan [
-|a|
-a:=1 + value squared.
-^(super arcTan) eps2: eps2 /  a; eps1eps2: eps1eps2 /a + (-2 * eps * eps2 * value /(a squared));yourself
-
+	| a |
+	a := 1 + value squared.
+	^ super arcTan
+		eps2: eps2 / a;
+		eps1eps2: eps1eps2 / a + (-2 * eps * eps2 * value / a squared);
+		yourself
 ]
 
 { #category : #'mathematical functions' }
 PMHyperDualNumber >> cos [
-^(super cos) eps2: eps2 * value sin negated; eps1eps2: (eps1eps2 * value sin + (eps * eps2 * value cos))negated ;yourself
-
+	^ super cos
+		eps2: eps2 * value sin negated;
+		eps1eps2: (eps1eps2 * value sin + (eps * eps2 * value cos)) negated;
+		yourself
 ]
 
 { #category : #accessing }
 PMHyperDualNumber >> eps1eps2 [
-	^eps1eps2
+	^ eps1eps2
 ]
 
 { #category : #accessing }
 PMHyperDualNumber >> eps1eps2: anEps [
-	eps1eps2 := anEps 
+	eps1eps2 := anEps
```

Comment:
A thought I had was that instances of `PMHyperDualNumber` should be immutable. I'd prefer we address it now. Do you feel the same? We could make it private to ensure clients never get to use it.

Reply:
You mean remove ``PMDualNumber>>eps:`` and ``PMHyperDualNumber>>eps1eps2:`` ? Make sense, some clients are using them at the moment.

Reply:
You mean, put ``PMDualNumber>>eps:`` and ``PMHyperDualNumber>>eps1eps2:`` in a private protocol ?

Reply:
Yes. I feel so, provided it is not being used by other client code.

Reply:
This is already used in some PMDualNumber and PMHyperDualNumber methods.

Reply:
Ah. Fair enough. Perhaps in some documentation we should advise against using this method, then because it allows numbers (value objects) to become mutable.

Reply:
We could use immutability and define them as beReadOnlyObject.

Reply:
That'd be perfect! :heart:

Reply:
I try to convert PMDualNumberand PMHyperDualNumber as classes that generate immutable objects. This is ok with PMDualNumber but a bit a pain for PMHyperDualNumber, because the last class is a subclass of the previous one. Then this is not possible to reuse operations with super, because of the immutability. So it means we have to duplicate code between the two classes. There is a kind of tradeoff with immutability and code reuse.

Reply:
I'd prefer immutability over code reuse. As I was taught by Sandi Metz, it's easier to recover from code duplicate than from bad abstraction. I think I need to understand what a dual and hyper dual number is. Inheritance shouldn't be applied to re-use code. Could there be a Liskov Substitution problem here?

Reply:
PMHyperDualNumber is a dual number with 2 epsilons values.
PMDualNumber is a dual number with only one epsilon value.
This is why I think these 2 classes will be supersede later by an implementation of multi-dimensional dual numbers.